[11.x] Added optional arguments to fullUrlWithoutQuery() and fullUrlWithQuery() method in Request and Arr::query() to allow custom URI encoding when using these methods.

benmAltDesign opened the pull request:
Came across the issue of needing a custom encoding type while working on a redirect addon. For the moments I've reimplemented these functions but I'd like it to be optional in the core framework.

Default behaviour should be preserved with the default values in the optional args

Arr.php - Added optional argument in query() method for encoding type passed to http_build_query to allow custom encoding types (Default it set to the type that was previously hard-coded)

Request.php - Added optional argument to the fullUrlWithoutQuery() method to allow a custom encoding type for the URI from Arr::query() default is configured as the previous hardcoded default inside Arr::query() so the default behaviour without the optional arg remains the same

<!--
Please only send a pull request to branches that are currently supported: https://laravel.com/docs/releases#support-policy 

If you are unsure which branch your pull request should be sent to, please read: https://laravel.com/docs/contributions#which-branch

Pull requests without a descriptive title, thorough description, or tests will be closed.

In addition, please describe the benefit to end users; the reasons it does not break any existing features; how it makes building web applications easier, etc.
-->

benmAltDesign committed:
Arr.php - Added optional argument in query() method for encoding type passed to http_build_query to allow custom encoding types (Default it set to the type that was previously hard-coded)

Request.php - Added optional argument to the fullUrlWithoutQuery() method to allow a custom encoding type for the URI from Arr::query() default is configured as the previous hardcoded default inside Arr::query() so the default behaviour without the optional arg remains the same

benmAltDesign committed:
Code styling fixes as flagged by styleci

shaedrich commented on src/Illuminate/Collections/Arr.php:
```diff
@@ -702,11 +702,12 @@ public static function pull(&$array, $key, $default = null)
      * @param  array  $array
+     * @param  int-mask $encoding_type (optional) PHP Query encoding type.
      * @return string
      */
-    public static function query($array)
+    public static function query($array, $encoding_type = PHP_QUERY_RFC3986)
     {
-        return http_build_query($array, '', '&', PHP_QUERY_RFC3986);
+        return http_build_query($array, '', '&', $encoding_type);
```
I'd use camel case here:
```suggestion
     * @param  int-mask $encodingType (optional) PHP Query encoding type.
     * @return string
     */
    public static function query($array, $encodingType = PHP_QUERY_RFC3986)
    {
        return http_build_query($array, '', '&', $encodingType);
```

benmAltDesign replied:
Switched it to camelCase in a newer commit

shaedrich commented on src/Illuminate/Collections/Arr.php:
```diff
@@ -702,11 +702,12 @@ public static function pull(&$array, $key, $default = null)
      * Convert the array into a query string.
      *
      * @param  array  $array
+     * @param  int-mask $encoding_type (optional) PHP Query encoding type.
```
Since we can narrow it down, we should do just that, to improve IDE support:
```suggestion
     * @param  int-mask-of<PHP_QUERY_*> $encoding_type (optional) PHP Query encoding type.
```

shaedrich commented on src/Illuminate/Collections/Arr.php:
```diff
@@ -702,11 +702,12 @@ public static function pull(&$array, $key, $default = null)
      * Convert the array into a query string.
      *
      * @param  array  $array
+     * @param  int-mask $encoding_type (optional) PHP Query encoding type.
```
Since we are in PHP, I'd say, we don't need to mention that
```suggestion
     * @param  int-mask $encoding_type (optional) Query encoding type.
```

benmAltDesign replied:
removed "PHP" from the @param in both PHPDoc comments

shaedrich commented on src/Illuminate/Collections/Arr.php:
```diff
@@ -702,11 +702,12 @@ public static function pull(&$array, $key, $default = null)
      * @param  array  $array
+     * @param  int-mask $encoding_type (optional) PHP Query encoding type.
      * @return string
      */
-    public static function query($array)
+    public static function query($array, $encoding_type = PHP_QUERY_RFC3986)
     {
-        return http_build_query($array, '', '&', PHP_QUERY_RFC3986);
+        return http_build_query($array, '', '&', $encoding_type);
```
We might also create an enum for this 🤔 
```suggestion
     * @param  int-mask|HttpQueryEncoding $encoding_type (optional) PHP Query encoding type.
     * @return string
     */
    public static function query($array, $encoding_type = HttpQueryEncoding::Rfc3986)
    {
        return http_build_query($array, '', '&', enum_value($encoding_type));
```

benmAltDesign committed:
Update src/Illuminate/Collections/Arr.php - Narrow down mask type

Co-authored-by: Sebastian Hädrich <11225821+shaedrich@users.noreply.github.com>

benmAltDesign committed:
Added suggested items from code review except the enum.

benmAltDesign commented:
@shaedrich Hi, I've added all the changed except your enum idea because I merged one in early by accident and I can't merge that in now due to it being classed as outdated. Could you suggest that change again and I'll pull that in, there's some great suggestions in there. Especially the int-mask-of in the PHPDoc

Apologies

benmAltDesign committed:
Remove PHP from PHP doc since it's obvious we're in PHP

shaedrich commented on src/Illuminate/Collections/Arr.php:
```diff
@@ -702,11 +702,12 @@ public static function pull(&$array, $key, $default = null)
      * Convert the array into a query string.
      *
      * @param  array  $array
+     * @param  int-mask-of<PHP_QUERY_*> $encodingType (optional) PHP Query encoding type.
```
Analogously to `Request::fullUrlWithoutQuery()` as requested in https://github.com/laravel/framework/pull/54464#discussion_r1941392172 and changed in 52b2123, the language would have to be removed here as well:
```suggestion
     * @param  int-mask-of<PHP_QUERY_*> $encodingType (optional) Query encoding type.
```

shaedrich commented:
@benmAltDesign Sure thing, yeah, sometimes, I'm surprised, what PHP is capable of in terms of static typing

benmAltDesign commented:
> @benmAltDesign Sure thing, yeah, sometimes, I'm surprised, what PHP is capable of in terms of static typing

Yeah it really is, I'm a bit rusty on my PHPDoc so I wasn't aware you could do that kind of type hint, though it makes perfect sense that the standard library functions etc must do it somehow

benmAltDesign committed:
Added Enum for HttpQueryEncoding

benmAltDesign commented:
@shaedrich I've added this in with one minor change. In the enum I've added RFC1738 at the top since it is defined as 1 and the query encoding it defined as 2, it makes more sense for them to be the other way around I think.

Cheers for suggestion

shaedrich commented:
@benmAltDesign Thanks a lot! Oh, good catch! 👍🏻  Yeah, that was an oversight on my part ^^

benmAltDesign committed:
Uri.php - Added a getRaw() method to the URI to allow getting the URI path without it being stripped of / characters.

Request.php - Updated fullUrlWithQuery to accept an optional encoding type argument to match previous changes to fullUrlWithoutQuery

Restructured fullUrlWithoutQuery and fullUrlWithQuery, moved the main logic to matching Uri versions of these functions. UriWithoutQuery and UriWithQuery are then used the get the Uri version which in the original fullUrl version has the schemeAndHttpHost() prepended to them.

 This should add the ability to use these Uri functions whilst maintaining the existing functionality in the existing functions and not duplicating the code.

benmAltDesign committed:
Reverted big changes that broke a load of test, just adding the enconding type functionality to the fullUrlWithQuery() method in Request now

taylorotwell commented:
Thanks for your pull request to Laravel!

Unfortunately, I'm going to delay merging this code for now. To preserve our ability to adequately maintain the framework, we need to be very careful regarding the amount of code we include. 

If applicable, please consider releasing your code as a package so that the community can still take advantage of your contributions!

benmAltDesign commented:
No problem, cheers for looking 🫡 

Thanks for the great work and awesome framework
